File: src/FXRateDashboard/Models/RateSeriesSnapshot.cs
namespace FXRateDashboard.Models;

public sealed class RateSeriesSnapshot
{
    public string Pair { get; set; } = string.Empty;

    public TimeRangePreset Range { get; set; }

    public List<RatePoint> Points { get; set; } = [];

    public decimal CurrentRate { get; set; }

    public decimal ChangeAbsolute { get; set; }

    public decimal ChangePercent { get; set; }

    public DateTimeOffset AsOfUtc { get; set; }

    public bool IsStale { get; set; }

    public RateSeriesSnapshot Clone()
    {
        return new RateSeriesSnapshot
        {
            Pair = Pair,
            Range = Range,
            Points = Points.Select(point => new RatePoint
            {
                TimestampUtc = point.TimestampUtc,
                Rate = point.Rate
            }).ToList(),
            CurrentRate = CurrentRate,
            ChangeAbsolute = ChangeAbsolute,
            ChangePercent = ChangePercent,
            AsOfUtc = AsOfUtc,
            IsStale = IsStale
        };
    }
}

File: macos/FXRateDashboard/Sources/FXRateDashboardKit/Models/RateSeriesSnapshot.swift
import Foundation

public struct RateSeriesSnapshot: Codable, Equatable, Sendable {
    public var pair: String
    public var range: TimeRangePreset
    public var points: [RatePoint]
    public var currentRate: Decimal
    public var changeAbsolute: Decimal
    public var changePercent: Decimal
    public var asOfUTC: Date
    public var isStale: Bool

    public init(
        pair: String,
        range: TimeRangePreset,
        points: [RatePoint],
        currentRate: Decimal,
        changeAbsolute: Decimal,
        changePercent: Decimal,
        asOfUTC: Date,
        isStale: Bool
    ) {
        self.pair = pair
        self.range = range
        self.points = points
        self.currentRate = currentRate
        self.changeAbsolute = changeAbsolute
        self.changePercent = changePercent
        self.asOfUTC = asOfUTC
        self.isStale = isStale
    }
}
